keystore/filesystem/server_keystore.go

```diff
@@ -1216,14 +1238,32 @@ func (store *KeyStore) generateAndSaveSymmetricKey(id []byte, filename string) e
 func (store *KeyStore) readEncryptedKey(id []byte, filename string) ([]byte, error) {
 	encryptedSymKey, ok := store.Get(filename)
 	if !ok {
-		var err error
-		encryptedSymKey, err = store.ReadKeyFile(store.GetPrivateKeyFilePath(filename))
-		if err != nil {
-			return nil, err
-		}
-		store.Add(filename, encryptedSymKey)
+		return store.loadKeyAndCache(id, filename, func() ([]byte, error) {
+			return store.ReadKeyFile(store.GetPrivateKeyFilePath(filename))
+		})
+	}
+	return store.cacheEncryptor.Decrypt(encryptedSymKey, id)
+}
+
+func (store *KeyStore) loadKeyAndCache(id []byte, filename string, loadKeyCallback func() ([]byte, error)) ([]byte, error) {
+	encryptedKey, err := loadKeyCallback()
```

Comment:
The idea with callback is cool, but I don't understand why not just pass the key directly? The callback is used immediately and only once, so I don't see any reasons to store it. Therefore, instead of this:
```go
return store.loadKeyAndCache(id, filename, func() ([]byte, error) {
    return store.ReadKeyFile(store.GetPrivateKeyFilePath(filename))
})
```

We can write something like:
```go
key, err := store.ReadKeyFile(store.GetPrivateKeyFilePath(filename))
if err != nil { ... }
return store.cacheKey(id, filename, key)
```

Though, at least callback requires less boilerplate code for error handling :)

Reply:
@G1gg1L3s, Yes, the idea was to create some compact function that does all the necessary logic under the hood.

Just thought about some sort of improvement. We can do the following:

```
loadKeyAndCache(id []byte, filename string, loadKeyCallback func(filePath string) ([]byte, error)) ([]byte, error) {
	encryptedKey, err := loadKeyCallback(store.GetPrivateKeyFilePath(filename))
	if err != nil {
		return nil, err
	}
	.....
	.....
```

And on the caller side just:

```
store.loadKeyAndCache([]byte(SecureLogKeyFilename), filename, store.ReadKeyFile)
```

But with such approach, we can lose some flexibility in future variants of getting the path to keys. What do you think guys? cc @Lagovas

Reply:
as I understand, we cannot use @G1gg1L3s's approach:
```
key, err := store.ReadKeyFile(store.GetPrivateKeyFilePath(filename))
if err != nil { ... }
return store.cacheKey(id, filename, key)
```
because this snippet returns symmetric key. But we also have case with asymmetric keys that uses `loadPrivateKey` function - https://github.com/cossacklabs/acra/blob/03f7f071f33ef344d920c08387c8cffd9d160b97/keystore/filesystem/server_keystore.go#L580

So we cannot use only one snippet, we should use several. And everywhere repeat flow:
* read encrypted key
* decrypt
* encrypt for cache
* store in cache
* return decrypted

And this callback approach helps encapsulate key decryption, encryption and saving in cache steps, and avoid extra row repeating. Or did I miss something?

Reply:
Yes, you are right, it was exactly the reason for using it.